### ISSUE TITLE
[ViacomCBS] - 4001 Error with Dash Manifests on Xbox 

### ISSUE THREAD

**Issue by agajassi** [google/shaka-player#3636]
**Have you read the [Tutorials](http://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html)?**
Yes I have read it

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes I've read FAQ and couldn't find relevant question. Also looked for duplicate open issues and didn't see any.

**What version of Shaka Player are you using?**
3.2.0

**Please ask your question**

On Xbox whenever I attempt to play any Dash manifest I get this following error:
`4001: MANIFEST.DASH_INVALID_XML`

This only occurs on Xbox and not other platforms. Also this does not happen with Shaka Player 3.0.13. Not sure if anything changed with later versions of the player, or there is a problem with the XML schema. I've tried this with few different Dash manifests, including the following test asset from Bitmovin:
https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd

This occurs on Shaka 3.2.0 prod release and I also tried master branch and can reproduce on both. I thought I would first ask  question, before I file a ticket. I searched and didn't find any duplicate tickets. Is this a known issue?

I tried this with very simple vanilla Shaka player configuration:
```
        const url = 'https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd';
        const video = document.querySelector('video');
        shaka.polyfill.installAll();
        shaka.log.setLevel(shaka.log.Level.DEBUG);
        const player = new shaka.Player(video);
        const config = player.getConfiguration();
        player.configure(config);

        player.load(url, 1500)
            .then((e) => {
                console.log('ready');
            })
            .catch((e) => {
                onError(e);
            });

        function onErrorEvent(event) {
            onError(event.detail);
        }

        function onError(error) {
            console.error('Error code', error.code, 'object', error);
        }
```

Some sort of guidance on this problem would be greatly appreciated.

![20210915_110708](https://user-images.githubusercontent.com/12893220/133497398-e22b9ea4-3a59-41e7-bef4-8aeb9356ac01.jpg)

**Comment by joeyparrish** [comment]
This is not a known issue, but we use the browser's built-in DOM parser to parse XML.  Either the platform won't parse the XML, or there is another exception being caught in the same code snippet as the parsing.

I don't have debugger access on an Xbox, but I can use our test runner to run automated snippets on the device and see what happens.  I will try parsing this manifest's content directly in DOMParser and see what I get.

**Comment by joeyparrish** [comment]
Via our test runner, I tried all of the following:

1. Downloading the manifest content and parsing it through DOMParser
2. Downloading the manifest content and parsing it through our XmlUtils abstraction that normally wraps DOMParser in our codebase
3. Passing the manifest URI to the complete DASH parser

All of these worked and successfully parsed the manifest on my Xbox One device.  The only difference I can find is that these did not involve the compiled library.  I'll try again with the compiled bundle and see if there is any difference.

**Comment by joeyparrish** [comment]
Wait, I found it.  I overlooked something in my previous tests.

The difference is that this MPD has a lowercase `<mpd>` element at its root, and we do a case-sensitive comparison checking the name of the root element (expecting `<MPD>`).  We have a provision that treats a mismatch on the root element as a parsing failure.

**Comment by joeyparrish** [comment]
No, I take it back.  In this content, **all** tags are lowercase.  It would take a lot of mess all over the DASH parser to tolerate this.

I'm going to label this as "bad content".  Please correct the generation of this manifest.

**Comment by joeyparrish** [comment]
We're all going to get whiplash here.  The debugger was printing the XML element objects in lowercase, even though the source isn't.

So... no idea what's wrong yet.  😢

**Comment by joeyparrish** [comment]
After correcting bugs in my test code, I am unable to reproduce this error on my Xbox One.

I tried:

1. Fetching the manifest and parsing it using DOMParser
2. Fetching the manifest and parsing it using our XmlUtils abstraction
3. Feeding the manifest to the uncompiled DASH parser
4. Feeding the manifest to the compiled DASH parser

All of these produced valid results, and no errors.

My device reports its User-Agent string as: `Mozilla/5.0 (Windows NT 10.0; Win64; x64; Xbox; Xbox One; WebView/3.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36 Edge/18.19041`

Here's my test code:

```js
describe('Error 4001', () => {
  it('ad-hoc test for issues 3636 and 3627', async () => {
    // This is the URI from #3636; the URI from #3627 is private.
    const uri =
        'https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd';
    const data = await shaka.test.Util.fetch(uri);
    const mimeType = 'text/xml';

    // Parse with DOMParser directly
    const dataString = shaka.util.StringUtils.fromUTF8(data);
    const parser = new DOMParser();
    const doc1 = parser.parseFromString(dataString, mimeType);
    console.log(doc1);
    if (doc1 == null) {
      fail('doc1 is null');
    }

    // Parse with our XmlUtils abstraction
    const doc2 = shaka.util.XmlUtils.parseXml(data, 'MPD');
    console.log(doc2);
    if (doc2 == null) {
      fail('doc2 is null');
    }

    const manifestConfig =
        shaka.util.PlayerConfiguration.createDefault().manifest;

    // Parse with uncompiled DASH parser
    const playerInterface1 = {
      networkingEngine: new shaka.net.NetworkingEngine(),
      filter: (manifest) => {},
      makeTextStreamsForClosedCaptions: (manifest) => {},
      onTimelineRegionAdded: (region) => {},
      onEvent: (event) => {},
      onError: (error) => fail(error),
      isLowLatencyMode: () => false,
      isAutoLowLatencyMode: () => false,
      enableLowLatencyMode: () => {},
    };
    const dashParser1 = new shaka.dash.DashParser();
    dashParser1.configure(manifestConfig);
    // Throws on error, failing the test automatically
    const manifest1 = await dashParser1.start(uri, playerInterface1);
    console.log(manifest1);

    const compiledShaka =
        await shaka.test.Loader.loadShaka(/* uncompiled= */ false);

    // Parse with compiled DASH parser
    const playerInterface2 = {
      networkingEngine: new compiledShaka.net.NetworkingEngine(),
      filter: (manifest) => {},
      makeTextStreamsForClosedCaptions: (manifest) => {},
      onTimelineRegionAdded: (region) => {},
      onEvent: (event) => {},
      onError: (error) => fail(error),
      isLowLatencyMode: () => false,
      isAutoLowLatencyMode: () => false,
      enableLowLatencyMode: () => {},
    };
    const dashParser2 = new compiledShaka.dash.DashParser();
    dashParser2.configure(manifestConfig);
    // Throws on error, failing the test automatically
    const manifest2 = await dashParser2.start(uri, playerInterface2);
    console.log(manifest2);
  });
});
```

**Comment by agajassi** [comment]
@joeyparrish 
Thank you. I was on a PTO, so my apologies for a late reply.
I've tried this issue once again. I am still able to reproduce it with Xbox One and Shaka Player v.3.2.0, so are my other 2 colleagues here at ViacomCBS. 

If I switch to Shaka Player v. 3.0.10, then it plays without any issues. I went ahead and tried other versions as well in attempt to pinpoint when this started breaking. It looks like last Shaka Player version where this is not reproducible is 3.0.13. After 3.0.13 and above we can reproduce. 

I can keep trying and investigating further, but if you have any ideas could you please let me know?

You mentioned that after correcting some bugs in your tests, you are able to get it to play. I am curious what is it that you did to resolve this.

**Comment by joeyparrish** [comment]
The test code above is everything you need.  I had bugs in an earlier version of it, specifically around my checks for failure in the first two steps (using DOMParser and XmlUtils to parse).  The test code above should be correct, though.

The test tries four things to reproduce the error, each at a different level of the stack:

1. Fetching the manifest and parsing it using DOMParser
2. Fetching the manifest and parsing it using our XmlUtils abstraction
3. Feeding the manifest to the uncompiled DASH parser
4. Feeding the manifest to the compiled DASH parser

If some of these fail and other pass, it will tell us at which level of the stack the bug lies.

I would love it if you could try running the same test on your device and letting me know the results.  (Feel free to update the URI we're parsing, though, to substitute the private URI you sent us.)  Let me know if you need help getting the automated tests running on the device.

**Comment by agajassi** [comment]
When I attempt to play an actual mpd content with 3.2.0 on Xbox, it looks like it attempts to parse the manifest using 2nd method (Dash Parser), because I get the same error

**Comment by joeyparrish** [comment]
Please try with a TextDecoder polyfill installed, as we do in our demo and tests.  This is a requirement for platforms without a native implementation, and not having it will cause `shaka.Player.isBrowserSupported()` to return `false`.  This could be the root of your problem if you don't have such a polyfill in your application on Xbox.

We install `fastestsmallesttextencoderdecoder` and load `node_modules/fastestsmallesttextencoderdecoder/EncoderDecoderTogether.min.js`.  There is a brief mention in our [v3 Upgrade Guide](https://shaka-player-demo.appspot.com/docs/api/tutorial-upgrade-v3.html), but if this was too difficult to find, we can put something more obvious into the README.

Does this help?  Does this change your test results?  Does this explain the issue in your application?
